ipv8/peerdiscovery/network.py

```diff
@@ -1,140 +1,175 @@
-        self.verified_peers = set()
+        self._all_addresses: Dict[Address, WalkableAddress] = {}
+        '''All known IP:port addresses, mapped to (introduction peer, services, new_style)'''
+
+        self.verified_peers: Set[Peer] = set()
+        '''All verified Peer objects (Peer.address must be in _all_addresses)'''
+
+        self.verified_by_public_key_bin: Dict[PublicKeyMat, Peer] = {}
+        '''Map of known public keys to Peer objects.'''
+
         self.graph_lock = RLock()
-        # Peers we should not add to the network
-        # For example, bootstrap peers
-        self.blacklist = []
-        # Excluded mids
-        self.blacklist_mids = []
-
-        # Map of advertised services (set) per peer
-        self.services_per_peer = {}
-        # Map of service identifiers to local overlays
-        self.service_overlays = {}
-
-        # Cache of IP:port -> Peer
-        self.reverse_ip_lookup = OrderedDict()
+        '''Lock for all updates of the peer pool.'''
+
+        self.blacklist: List[Address] = []
+        '''Peers we should not add to the network (e.g., bootstrap peers), by address.'''
+
+        self.blacklist_mids: List[MID] = []
+        '''Peers we should not add to the network (e.g., bootstrap peers), by mid.'''
+
+        self.services_per_peer: Dict[PublicKeyMat, ServiceSet] = {}
+        '''Map of advertised services (set) per peer'''
+
+        self.service_overlays: Dict[Service, Overlay] = {}
+        '''Map of service identifiers to local overlays'''
+
         self.reverse_ip_cache_size = 500
+        self.reverse_ip_lookup = OrderedDict()
+        '''Cache of IP:port -> Peer. This is not a normal dict, as the amount of addresses can grow infinitely!'''
 
-        # Cache of Peer -> [IP:port]
-        self.reverse_intro_lookup = OrderedDict()
         self.reverse_intro_cache_size = 500
+        self.reverse_intro_lookup = OrderedDict()
+        '''Map of Peer -> [IP:port], reversing the information from _all_addresses.
+        This is not a normal dict, as the amount of addresses can grow infinitely!'''
 
-        # Cache of service_id -> [Peer]
-        self.reverse_service_lookup = OrderedDict()
         self.reverse_service_cache_size = 500
+        self.reverse_service_lookup = OrderedDict()
+        '''Cache of service_id -> [Peer]. This is not a normal dict, as the amount of services can grow infinitely!'''
 
-    def is_new_style(self, address):
-        return self._all_addresses.get(address, ('', None, False))[2]
+    def is_new_style(self, address: Address) -> bool:
+        """
+        Check if an address supports new-style introduction requests and responses.
+
+        :param address: the address to check for.
+        :returns: True iff the address is both known and known to support new-style introductions.
+        """
+        return self._all_addresses.get(address, WalkableAddress(b'', None, False)).new_style
```

Comment:
The same here, it is possible to avoid the creation of WlakableAddress if it is already exists:
```python
return (self._all_addresses.get(address) or WalkableAddress(b'', None, False)).new_style
```